ttnn/cpp/ttnn/operations/data_movement/slice/slice.cpp

```diff
@@ -8,133 +8,239 @@
 #include "device/slice_op.hpp"
 #include "ttnn/run_operation.hpp"
 #include "ttnn/operations/core/core.hpp"
+#include "ttnn/cpp/ttnn/operations/creation.hpp"
 
 
 namespace ttnn::operations::data_movement {
+namespace detail {
+    uint32_t wrap_index(int index, int size) {
+        return index < 0 ? size + index : index;
+    }
+    uint32_t round_up_to_multiple_of_32(uint32_t value) {
+        return value == 0 ? 32 : ((value + 31) & ~31);
+    }
+}
 
+template<typename T>
 ttnn::Tensor SliceOperation::invoke(
     uint8_t queue_id,
     const ttnn::Tensor& input_tensor,
-    tt::tt_metal::LegacyShape output_tensor_start,
-    tt::tt_metal::LegacyShape output_tensor_end,
-    const std::optional<tt::tt_metal::LegacyShape> step,
+    const std::vector<T> &begins,
+    const std::vector<T> &ends,
+    const std::vector<T> &step,
     const std::optional<MemoryConfig>& memory_config_arg) {
-    std::optional<tt::tt_metal::LegacyShape> modified_step = step;
-    if (modified_step.has_value()) {
-        if (std::all_of(modified_step->begin(), modified_step->end(), [](int32_t s) { return s == 1; })) {
-            modified_step = std::nullopt;
+
+    // Ensure start and end vectors have matching sizes and correct tensor rank
+    uint32_t input_rank = input_tensor.get_shape().rank();
+    TT_FATAL(input_rank == begins.size(), "Start and end must have the same size");
```

Comment:
wrong error description

also, can you add details to error messages, like actual values?